### PR TITLE
Refactor the Automated Test Runner, and add retry logic

### PR DESCRIPTION
**Pull request by a-maurice**
Refactor the automated test runner to treat the tests as objects instead of just functions.  Add retry logic, either set for the either set of tests, or on an individual test basis.
***

Full test run: https://github.com/firebase/firebase-unity-sdk/actions/runs/3962008250
Tests that use the retry logic: https://github.com/firebase/firebase-unity-sdk/actions/runs/3954504632/jobs/6772138771
***

### Type of Change
Place an `x` the applicable box:
- [ ] Bug fix. Add the issue # below if applicable.
- [ ] New feature. A non-breaking change which adds functionality.
- [x] Other, such as a build process or documentation change.
***

### CODE REVIEW

File: scripts/gha/integration_testing/automated_testapp/AutomatedTestRunner.cs [hunk 1]
```diff
@@ -244,4 +215,190 @@ public class MainThreadDispatcher : MonoBehaviour {
...
+                                  testExceptions[i].ToString()));
+          }
+        }
+      } else if (Status == AutomatedTestStatus.TimedOut) {
+        LogFunc("Test " + Description + " timed out, elapsed time: " + timeOutTime);
+      }
+    }
+
+    public void CollectWarnings(List<string> warnings) {
+      // If multiple attempts were taked, add a warning
```
Comment (DellaBitta) [inline scripts/gha/integration_testing/automated_testapp/AutomatedTestRunner.cs:388]: nit: Add a warning if multiple attempts were made.

File: scripts/gha/integration_testing/automated_testapp/AutomatedTestRunner.cs [hunk 2]
```diff
@@ -244,4 +215,190 @@ public class MainThreadDispatcher : MonoBehaviour {
...
+      if (TookMultipleAttempts) {
+        warnings.Add(String.Format("Test {0} took {1} of {2} attempts to pass.",
```
Comment (DellaBitta) [inline scripts/gha/integration_testing/automated_testapp/AutomatedTestRunner.cs:390]: Should we add the word Flake somewhere in this message?

File: scripts/gha/integration_testing/automated_testapp/AutomatedTestRunner.cs [hunk 3]
```diff
@@ -46,27 +52,38 @@ public class AutomatedTestRunner {
...
-      customTestNames = testNames;
       Finished = false;
       TestTimeoutSeconds = 60.0f;
       this.testLabManager = testLabManager;
+      tests = new List<AutomatedTestCase>();
+
+      for (int i = 0; i < testsToRun.Length; ++i) {
+        string testName = null;
+        if (testNames != null && i < testNames.Length) {
+          testName = testNames[i];
```
[the remark this answers is not in the document]
Comment (a-maurice) [reply scripts/gha/integration_testing/automated_testapp/AutomatedTestRunner.cs:66]: There is no guarantee that testNames exists, or is the same length as testsToRun.  So I added the extra parts to be a bit safer.

File: scripts/gha/integration_testing/automated_testapp/AutomatedTestRunner.cs [hunk 4]
```diff
@@ -9,23 +9,29 @@ namespace Firebase.Sample {
...
-    private List<Exception> testExceptions = new List<Exception>();
-    // Whether the currently running test is expected to fail.
-    private bool postTestIgnoredFailureCheckEnabled = true;
-    // Name and index of the current running test function.
-    public string CurrentTestDescription { get; private set; }
-    // The Task that represents the currently running test.
-    public Task CurrentTestResult { get; private set; }
+    private bool startedExecution = false;
+
+    private List<AutomatedTestCase> tests = null;
```
[the remark this answers is not in the document]
Comment (a-maurice) [reply scripts/gha/integration_testing/automated_testapp/AutomatedTestRunner.cs:16]: Added it back

File: scripts/gha/integration_testing/automated_testapp/AutomatedTestRunner.cs (hunk 1 above)
Comment (a-maurice) [reply scripts/gha/integration_testing/automated_testapp/AutomatedTestRunner.cs:388]: Done

File: scripts/gha/integration_testing/automated_testapp/AutomatedTestRunner.cs (hunk 2 above)
Comment (a-maurice) [reply scripts/gha/integration_testing/automated_testapp/AutomatedTestRunner.cs:390]: Done

File: scripts/gha/integration_testing/automated_testapp/AutomatedTestRunner.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (DellaBitta) [reply scripts/gha/integration_testing/automated_testapp/AutomatedTestRunner.cs:66]: Right, the if statement is great, but why not just do:

```
 if (testNames != null && i < testNames.Length) {
          AddTests(testNames[i]);
```

**Review by DellaBitta — APPROVED** [review]

**Pull request merged**